### DISCUSSION TITLE
machine.wake_reason()

### DISCUSSION THREAD

**Discussion by giancarlo64** [micropython/micropython#13314] [ESP32]
Hi,
I have to detect the cause of a wake-up of the ESP32 from deep-sleep. In my case It could be from a timer or from an external pin. 
For this reason  I use the fuction  machine.wake_reason() to detect it. 
First of all some description tell that the return value of this function should be a tuple other a constant. What is right?
Second: from the constants listed in micropython.org  (machine.WLAN_WAKE, machine.PIN_WAKE, machine.RTC_WAKE) only PIN.wake is recognized. The other constants during exection give me an error ('module' object has no attribute 'RTC_WAKE' .... )  
Third: I get a retured value from a timer deepsleep of  4 (the cause listed in the doc are 3) 

I'm a little bit confused how to use this function

Gian Carlo

**Comment by mattytrentini** [comment]
I haven't used this particular feature but I believe the implementation is in [modmachine.c:238](https://github.com/micropython/micropython/blob/master/ports/esp32/modmachine.c#L238). 

From what I can tell it appears to be converting the enumeration returned by the ESP-IDF call `esp_sleep_get_wakeup_cause()`. That enumeration is defined by [esp_sleep_source_t](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html#_CPPv418esp_sleep_source_t). 

There is currently no mapping between the values returned and MicroPython constants. Until such a mapping is put in place, I think it would be best to trigger the various wake-up causes and check the integer values returned. Then you can use those integer values to check the wakeup reason. Be aware they may change between ESP-IDF updates!

**Comment by giancarlo64** [comment]
Hi Mattytrentini,
thanks! I will do so.
